File: windows/src/Edgepad/Protocol/Frame.cs
namespace Edgepad.Protocol;

/// <summary>One message on the wire. Type bytes and payload layouts live in <see cref="FrameCodec"/>.</summary>
internal abstract record Frame;

internal sealed record Hello(byte Version) : Frame;

internal sealed record HelloAck(byte Version) : Frame;

internal sealed record Move(short Dx, short Dy) : Frame;

internal sealed record PointerButton(byte Id, bool Down) : Frame;

internal sealed record Scroll(short Dx, short Dy) : Frame;

internal sealed record Zoom(short Delta) : Frame;

internal sealed record RunAction(byte Id) : Frame;

internal sealed record SetValue(byte Control, byte Value) : Frame;

internal sealed record Ping(long Time) : Frame;

internal sealed record Pong(long Time) : Frame;

internal sealed record StateReport(byte Control, byte Value, byte Flags) : Frame;

/// <summary>Text, of a <see cref="TextKind"/>: what is playing, the app, the timeline, or characters to type. At most 255 UTF-8 bytes.</summary>
internal sealed record Text(byte Kind, string Value) : Frame;

/// <summary>One keyboard key, by Windows virtual-key code (e.g. 0x41 = A, 0x10 = SHIFT).</summary>
internal sealed record Key(ushort Code, bool Down) : Frame;

/// <summary>
/// The whole controller in one frame: every button, both triggers and both sticks together, sent when any of
/// them changes, and held here until the next one arrives. One frame rather than a frame per control because
/// a gamepad is read as a snapshot, not as a stream of edges — a dropped button release would otherwise stick
/// a button down until the user pressed it again.
/// <para>
/// The fields are deliberately byte-for-byte those of <c>XINPUT_GAMEPAD</c> (<c>wButtons</c>,
/// <c>bLeftTrigger</c>, <c>bRightTrigger</c>, <c>sThumbLX</c>, <c>sThumbLY</c>, <c>sThumbRX</c>,
/// <c>sThumbRY</c>), in that order and at those widths, so this side copies the payload into the struct
/// instead of translating it. A range or a button order of our own would be a conversion on the hot path and
/// a second definition to keep in step with Microsoft's.
/// </para>
/// <para>
/// <c>Buttons</c> is the <see cref="PadButton"/> mask. The triggers are 0..255. The four stick
/// axes are -32768..32767, positive up and right, exactly as XInput reports them.
/// </para>
/// </summary>
internal sealed record PadState(ushort Buttons, byte Lt, byte Rt, short Lx, short Ly, short Rx, short Ry) : Frame;

File: android/app/src/main/java/me/akshitbansal/edgepad/protocol/Frame.kt
package me.akshitbansal.edgepad.protocol

/** One message on the wire. Type bytes and payload layouts live in [FrameCodec]. */
sealed interface Frame {
    data class Hello(
        val version: Int,
    ) : Frame

    data class HelloAck(
        val version: Int,
    ) : Frame

    data class Move(
        val dx: Int,
        val dy: Int,
    ) : Frame

    data class PointerButton(
        val id: Int,
        val down: Boolean,
    ) : Frame

    data class Scroll(
        val dx: Int,
        val dy: Int,
    ) : Frame

    data class Zoom(
        val delta: Int,
    ) : Frame

    data class RunAction(
        val id: Int,
    ) : Frame

    data class SetValue(
        val control: Int,
        val value: Int,
    ) : Frame

    data class Ping(
        val time: Long,
    ) : Frame

    data class Pong(
        val time: Long,
    ) : Frame

    data class StateReport(
        val control: Int,
        val value: Int,
        val flags: Int,
    ) : Frame

    /** Text, of a [TextKind]: what is playing, the app, the timeline, or characters to type. At most 255 UTF-8 bytes. */
    data class Text(
        val kind: Int,
        val text: String,
    ) : Frame

    /** One keyboard key, by Windows virtual-key code (e.g. 0x41 = A, 0x10 = SHIFT). */
    data class Key(
        val code: Int,
        val down: Boolean,
    ) : Frame

    /**
     * The whole controller in one frame: every button, both triggers and both sticks together, sent when any
     * of them changes, with the laptop holding the last one until the next arrives. One frame rather than a
     * frame per control because a gamepad is read as a snapshot, not as a stream of edges — a dropped button
     * release would otherwise stick a button down until the user pressed it again.
     *
     * The field order and widths are deliberately byte-for-byte those of Windows' `XINPUT_GAMEPAD`
     * (`wButtons`, `bLeftTrigger`, `bRightTrigger`, `sThumbLX`, `sThumbLY`, `sThumbRX`, `sThumbRY`), so the
     * laptop copies the payload into the struct instead of translating it. Anything this side invented —
     * a different stick range, a button order of our own — would be a conversion on the hot path and a
     * second definition to keep in step with Microsoft's.
     *
     * [buttons] is XInput's `wButtons` mask, see [PadButton]. [lt] and [rt] are 0..255. The four stick axes
     * are -32768..32767, positive up and right, exactly as XInput reports them.
     */
    data class PadState(
        val buttons: Int,
        val lt: Int,
        val rt: Int,
        val lx: Int,
        val ly: Int,
        val rx: Int,
        val ry: Int,
    ) : Frame
}
